### DISCUSSION TITLE
pip-compile does not seem to respect python constraints

### DISCUSSION THREAD

**Discussion by nathanpower** [renovatebot/renovate#14011] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

31.66.3

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

I have the following config:

```JSON
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "prConcurrentLimit": 10,
  "constraints": {
    "python": "3.9"
  },
  "pip-compile": {
    "stabilityDays": 14,
    "prCreation": "not-pending",
    "labels": ["dependencies", "python"],
    "fileMatch": ["requirements-for-prophet.in", "requirements.in", "requirements-dev.in"],
    "packageRules": [],
    "vulnerabilityAlerts": {
      "labels": ["security", "python"],
      "stabilityDays": 0
    }
  },
  "docker": {
    "enabled": false
  },
  "docker-compose": {
    "enabled": false
  },
  "circleci": {
    "enabled": false
  },
  "pip_requirements": {
    "enabled": false
  },
  "js": {
    "enabled": false
  }
}
```

Upon running I get an error, which occurs because the command appears to be using python 3.10

I have also tried wrapping `constraints` inside `force` as per docs, same behaviour.

### Relevant debug logs

<details><summary>Logs</summary>

```
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "pip-compile requirements.in",
         "stdout": "",
         "stderr": "Traceback (most recent call last):\n  File \"/usr/local/python/3.10.2/bin/pip-compile\", line 8, in <module>\n    sys.exit(cli())\n  ...)"
       }
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment]
How are you running renovate? Python constraints are only respected on `binarySource=docker`!

**Comment by nathanpower** [comment]
> How are you running renovate? Python constraints are only respected on binarySource=docker!

So I have got these results during both:

- Running locally via `docker run renovate/renovate`
- Running on CircleCI via [daniel-shuy/renovate](https://circleci.com/developer/orbs/orb/daniel-shuy/renovate) orb which just wraps the official docker image as I understand

**Reply by viceice** [reply]
If you need to follow contraints, you need to use docker mode (use the slim docker image which is configured for docker mode)

https://docs.renovatebot.com/self-hosted-configuration/#binarysource
https://github.com/renovatebot/docker-renovate#samples
